tests/test_http_client.py

```diff
@@ -993,3 +1014,342 @@ def test_encode_array(self):
+    def check_call(self, request_mock, mocker):
+        def check_call(
+            mock,
+            method,
+            url,
+            post_data,
+            headers,
+            is_streaming=False,
+            timeout=80,
+            times=None,
+        ):
+            times = times or 1
+            args = (method, url)
+            kwargs = {
+                "headers": headers,
+                "data": post_data,
+                "timeout": timeout,
+                "proxies": {"http": "http://slap/", "https": "http://slap/"},
+            }
+
+            if is_streaming:
+                kwargs["stream"] = True
+
+            calls = [mocker.call(*args, **kwargs) for _ in range(times)]
+            request_mock.AsyncClient().request.assert_has_calls(calls)
+
+        return check_call
+
+    async def make_request_async(
+        self, method, url, headers, post_data, timeout=80
+    ):
+        client = self.REQUEST_CLIENT(
+            verify_ssl_certs=True, proxy="http://slap/", timeout=timeout
+        )
+        return await client.request_with_retries_async(
+            method, url, headers, post_data
+        )
+
+    async def make_request_stream_async(
+        self, method, url, headers, post_data, timeout=80
+    ):
+        client = self.REQUEST_CLIENT(
+            verify_ssl_certs=True, proxy="http://slap/"
+        )
+        return await client.request_stream_with_retries_async(
+            method, url, headers, post_data
+        )
+
+    @pytest.mark.asyncio
+    async def test_request(self, request_mock, mock_response, check_call):
+
+        mock_response('{"foo": "baz"}', 200)
+
+        for method in VALID_API_METHODS:
+            abs_url = self.valid_url
+            data = {}
+
+            if method != "post":
+                abs_url = "%s?%s" % (abs_url, data)
+                data = {}
+
+            headers = {"my-header": "header val"}
+            body, code, _ = await self.make_request_async(
+                method, abs_url, headers, data
+            )
+            assert code == 200
+            assert body == '{"foo": "baz"}'
+
+            check_call(request_mock, method, abs_url, data, headers)
+
+    @pytest.mark.asyncio
+    async def test_request_stream(
+        self, mocker, request_mock, mock_response, check_call
+    ):
+        # TODO
+        pass
+
+    async def test_exception(self, request_mock, mock_error):
+        mock_error(request_mock)
+        with pytest.raises(stripe.APIConnectionError):
+            await self.make_request_async("get", self.valid_url, {}, None)
+
+    @pytest.mark.asyncio
+    async def test_timeout(self, request_mock, mock_response, check_call):
+        headers = {"my-header": "header val"}
+        data = {}
+        mock_response('{"foo": "baz"}', 200)
+        await self.make_request_async(
+            "POST", self.valid_url, headers, data, timeout=5
+        )
+
+        check_call(None, "POST", self.valid_url, data, headers, timeout=5)
```

Comment:
This test method looks like it can be in the base test class as well.

Reply:
It isn't for the non-async clients. Timeout isn't in the base HTTPClient class so I think it shouldn't be in the base test, either.

Reply:
Do all async clients we are planning for support timeout? Maybe timeout should be on the base client if it's supported everywhere.

Reply:
What is the advantage of including it on the base client?

Reply:
1. We can expose timeout setting on client options and silently create an appropriate client in the background knowing that all our clients support timeout
2. We won't have to copy-paste timeout tests for every client and then accidentally miss one

Reply:
Not all sync clients support timeout, and I have a slight dislike for inconsistency between the async world and sync world here.

I'm also afraid that there will be subtle differences between what "timeout" means from http client to http client. Also in terms of missing things, there's the possibility that we will add another async client, where "timeout" is settable on the base client but we forget to pipe "timeout" through to the underlying client so the client will appear to support timeout when it doesn't.

It feels more conservative to me to keep timeout client-specific. I don't think this is a trapdoor, once more async clients are added we can look to see if timeout actually does work the same between them and choose to put them in the base class then. I would prefer that to trying to make a forecast now.

Reply:
>I'm also afraid that there will be subtle differences between what "timeout" means from http client to http client.
The base test will help catch these differences.

Can you file an issue for this? We do have a requirement for timeout option to be exposed in client options. So we'll need to solve this somehow.

Reply:
> Overall, I'd say that our sync client timeout implementation is not something to draw inspiration from.

Ha, very true.

I will file an issue -- I like the idea of using the AsyncClient as an opportunity to get timeouts right, but this is better to tackle separately.